mono/metadata/mono-perfcounters.c

```diff
@@ -1284,73 +1313,83 @@ custom_get_value_address (SharedCounter *scounter, SharedInstance* sinst)
+	SharedCounter * const scounter = find_custom_counter (cat, counter);
 	if (!scounter)
 		return NULL;
-	name = mono_string_to_utf8_checked (counter, error);
+
+	char * const name = mono_unwrapped_string_to_utf8 (counter, error);
 	return_val_if_nok (error, NULL);
+
 	*type = simple_type_to_type [scounter->type];
-	inst = custom_get_instance (cat, scounter, name);
+
+	SharedInstance* const inst = custom_get_instance (cat, scounter, name);
 	g_free (name);
 	if (!inst)
 		return NULL;
 	return custom_vtable (scounter, inst, (char *)custom_get_value_address (scounter, inst));
 }
 
 static const CategoryDesc*
-find_category (MonoString *category)
+find_category (mono_unwrapped_string_t category)
 {
 	int i;
-	for (i = 0; i < NUM_CATEGORIES; ++i) {
-		if (mono_string_compare_ascii (category, predef_categories [i].name) == 0)
-			return &predef_categories [i];
-	}
+	const CategoryDesc* desc = predef_categories;
+	for (i = 0; i < NUM_CATEGORIES; ++i, ++desc)
+		if (perfcounter_utf16_equal_ascii (category.chars, category.length, desc->name, desc->name_length))
+			return desc;
 	return NULL;
 }
 
 void*
-mono_perfcounter_get_impl (MonoString* category, MonoString* counter, MonoString* instance,
-		MonoString* machine, int *type, MonoBoolean *custom)
+ves_icall_System_Diagnostics_PerformanceCounter_GetImpl (
+		MonoStringHandle category, MonoStringHandle counter, MonoStringHandle instance,
+		MonoStringHandle machine, int *type, MonoBoolean *custom)
 {
+	HANDLE_FUNCTION_ENTER ()
+
 	MonoError error;
-	const CategoryDesc *cdesc;
+	error_init (&error);
+
 	void *result = NULL;
+	gchar *c_instance = NULL;
+	mono_unwrapped_string_t ucategory = mono_unwrap_string_handle (category);
+	mono_unwrapped_string_t ucounter = mono_unwrap_string_handle (counter);
+	mono_unwrapped_string_t uinstance = mono_unwrap_string_handle (instance);
+
 	/* no support for counters on other machines */
-	if (mono_string_compare_ascii (machine, "."))
-		return NULL;
-	cdesc = find_category (category);
+	if (!perfcounter_string_handle_equal_ascii (machine, ".", 1))
+		goto Null;
+
+	const CategoryDesc *cdesc;
+	cdesc = find_category (ucategory);
 	if (!cdesc) {
-		SharedCategory *scat = find_custom_category (category);
+		SharedCategory *scat = find_custom_category (ucategory);
 		if (!scat)
-			return NULL;
+			goto Null;
 		*custom = TRUE;
-		result = custom_get_impl (scat, counter, instance, type, &error);
+		result = custom_get_impl (scat, ucounter, uinstance, type, &error);
 		if (mono_error_set_pending_exception (&error))
-			return NULL;
-		return result;
+			goto Null;
+		goto Exit;
 	}
-	gchar *c_instance = mono_string_to_utf8_checked (instance, &error);
+	c_instance = mono_unwrapped_string_to_utf8 (uinstance, &error);
 	if (mono_error_set_pending_exception (&error))
-		return NULL;
+		goto Null;
 	switch (cdesc->id) {
 	case CATEGORY_CPU:
-		result = cpu_get_impl (counter, c_instance, type, custom);
+		result = cpu_get_impl (ucounter, c_instance, type, custom);
```

Comment:
Purely as a matter of personal preference: I try to arrange for the variable names not to change too much so that these -1/+1 diffs aren't needed in the course of converting code to use coop handles - makes it a bit clearer in git history that the behavior isn't changing.